docs/tutorials/forum-application/7.vote-extension.md

```diff
@@ -0,0 +1,125 @@
+
+	return &abci.ExtendVoteResponse{VoteExtension: []byte(app.CurseWords)}, nil
+}
+```
+
+**Explanation of code:**
+
+`ExtendVote` function takes two parameters: a `context.Context` and a pointer to an abci.`ExtendVoteRequest` struct.
+It returns a pointer to an `ExtendVoteResponse` struct and an error.
+
+The method implementation simply returns a new instance of `ExtendVoteResponse` with the `VoteExtension` field set to
+the value of `app.CurseWords.` The `app.CurseWords` is expected to be a byte array containing the vote extension data.
+
+The `ExtendVoteResponse` struct is used to encapsulate the response data for the `ExtendVote` method. By setting
+the `VoteExtension` field, the method includes the application-specific vote extension data in the response.
+
+In this implementation, the `ExtendVote` method in the ForumApp application returns the application-specific `vote extension`
+data stored in the `app.CurseWords` variable.
+
+*Tip: The `vote extensions` are opaque to the consensus algorithm but visible to the application, allowing for a variety
+of use cases like **price oracles, encrypted mempools, and threshold cryptography.***
+
+## VerifyVoteExtension
+
+The `VerifyVoteExtension` method allows applications to verify the `VoteExtension` data attached to each `pre-commit` message.
+
+When a validator is preparing to send a `pre-commit` vote, it first calls `ExtendVote` to generate a `VoteExtension.`
+This `VoteExtension` is broadcast along with the `pre-commit` vote.
+
+Other validators also call `ExtendVote` to generate their own `vote extensions.` However, not all validators will generate
+the same `vote extension.`
+
+When a validator receives a `pre-commit` vote with an attached `vote extension`, it calls `VerifyVoteExtension` to
+validate the `vote extension.`
+
+If the vote extension is successfully verified, the `pre-commit` vote is included in the tally. If validation fails,
+the entire `pre-commit` message is ignored.
+
+**Following is the blurb of code for the `VerifyVoteExtension` function:**
+
+```go
+// VerifyVoteExtension verifies the vote extensions and ensure they include the curse words
+// It will not be called for extensions generated by this validator
+func (app *ForumApp) VerifyVoteExtension(_ context.Context, req *abci.VerifyVoteExtensionRequest) (*abci.VerifyVoteExtensionResponse, error) {
+	fmt.Println("Executing Application VerifyVoteExtension")
+
+	if _, ok := app.valAddrToPubKeyMap[string(req.ValidatorAddress)]; !ok {
+		// we do not have a validator with this address mapped; this should never happen
+		panic(fmt.Errorf("unknown validator"))
+	}
+	curseWords := strings.Split(string(req.VoteExtension), "|")
+	tmpCurseWordMap := make(map[string]struct{})
+	// Verify that we do not have double words and the validator is not trying to cheat us
+	for _, word := range curseWords {
+		tmpCurseWordMap[word] = struct{}{}
+	}
+	if len(tmpCurseWordMap) < len(curseWords) {
```

Comment:
You might also want to add a limit to the number of words that fit in a vote extensions (for demonstration purposes). For a tutorial, it ca be hard-coded... or it could be set via a special ("gov-gated") transaction, but that'd be too much for a tutorial

Reply:
good idea, I've added code for that.